examples/pytorch/compgcn/model.py

```diff
@@ -0,0 +1,371 @@
+                 comp_fn='sub',
+                 activation=None,
+                 bias=True,
+                 dropout=0.0,
+                 batchnorm=False):
+        super(CompGraphConv, self).__init__()
+        self.hid_dim = hid_dim
+        self.comp_fn = comp_fn
+        self.activation = activation
+        self.bias = bias
+        self.dropout = nn.Dropout(dropout)
+        self.batchnorm = batchnorm
+
+        # define weights of 3 node matrics
+        self.W_I = nn.Parameter(th.Tensor(self.hid_dim, self.hid_dim))
+        self.W_O = nn.Parameter(th.Tensor(self.hid_dim, self.hid_dim))
+        self.W_S = nn.Parameter(th.Tensor(self.hid_dim, self.hid_dim))
+
+        if self.bias:
+            self.b_I = nn.Parameter(th.Tensor(self.hid_dim))
+            self.b_O = nn.Parameter(th.Tensor(self.hid_dim))
+            self.b_S = nn.Parameter(th.Tensor(self.hid_dim))
+
+        # define weights of >0 edge matric
+        self.W_rel = nn.Parameter(th.Tensor(self.hid_dim, self.hid_dim))
+        self.b_rel = nn.Parameter(th.Tensor(self.hid_dim))
+
+        # initialize all weights
+        self._initialize()
+
+    def _initialize(self):
+        r"""
+        Initialize all weights
+        """
+        nn.init.xavier_uniform_(self.W_I, gain=nn.init.calculate_gain('relu'))
+        nn.init.xavier_uniform_(self.W_O, gain=nn.init.calculate_gain('relu'))
+        nn.init.xavier_uniform_(self.W_S, gain=nn.init.calculate_gain('relu'))
+        nn.init.xavier_uniform_(self.W_rel, gain=nn.init.calculate_gain('relu'))
+        nn.init.zeros_(self.b_I)
+        nn.init.zeros_(self.b_O)
+        nn.init.zeros_(self.b_S)
+        nn.init.zeros_(self.b_rel)
+
+    def forward(self, g, n_in_feats, e_in_feats):
+        """
+        Input graph g should include reversed edges, and at the end of
+        forward add self loop to each node
+
+        Notes
+        -----
+        1. Node feature should have the same dimensions as edges so that
+        composition function could apply to them.
+        2. In CompGraphConv, the module assume the input graph has been
+        added reversed edge for each original edge. For example, if there
+        is a ('u' 'e' 'v') edge in the original graph, there must be a
+        ('v' 'e_inv' 'u') edge, where e_inv is mannual added to the original
+        graph AND edges in this type must have _inv in the end of edge name.
+
+        Parameters
+        ----------
+        g: DGL Graph
+            A DGL heterograph.
+
+        n_in_feats: dict[str, Tensor]
+            A dictionary of node type and node features.
+
+        e_in_feats: dict[str, Tensor]
+            A dictionary of edge type and edge features.
+
+        """
+        with g.local_scope():
+
+            # handle each edge type relation sub_graph
+            n_outputs = {ntype: [] for ntype in g.ntypes}
+
+            for stype, etype, dtype in g.canonical_etypes:
+
+                if etype not in e_in_feats:
+                    continue
```

Comment:
why this can happen?

Reply:
You are right. I follow this convention in RGCN. But this should not happen. All edges will use embeddings from basis vectors.

Reply:
Will remove this code.